File: 253_Meeting_Rooms_II/253.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e


class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        from Queue import PriorityQueue
        # O(nlgn)
        Q = PriorityQueue()
        intervals.sort(key=lambda itm: itm.start)
        for i, interval in enumerate(intervals):
            if Q.empty():
                Q.put(interval.end)
            else:
                head = Q.get()
                if head <= interval.start:
                    Q.put(interval.end)
                else:
                    Q.put(interval.end)
                    Q.put(head)
        return Q.qsize()

File: 253_Meeting_Rooms_II/253.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        priority_queue<int, std::vector<int>, std::greater<int>> last;
        sort(intervals.begin(), intervals.end(),
            [](Interval a, Interval b){return a.start < b.start;});
        for(int i = 0; i < intervals.size(); ++i){
            if(last.empty()){
                last.push(intervals[i].end);
            }else{
                cout<<last.top()<<endl;
                if(last.top() <= intervals[i].start){
                    last.pop();
                }
                last.push(intervals[i].end);
            }
        }
        return last.size();
    }
};